ir/translation/match_.rs

```diff
@@ -35,14 +35,26 @@ pub(crate) fn add_patterns(
         typeql::Pattern::Conjunction(nested) => add_patterns(function_index, conjunction, &nested.patterns),
         typeql::Pattern::Disjunction(disjunction) => add_disjunction(function_index, conjunction, disjunction),
         typeql::Pattern::Negation(negation) => add_negation(function_index, conjunction, negation),
-        typeql::Pattern::Optional(optional) => {
-            // add_optional(function_index, conjunction, optional)
-            Err(Box::new(RepresentationError::UnimplementedLanguageFeature {
-                feature: error::UnimplementedFeature::Optionals,
-            }))
-        }
+        typeql::Pattern::Optional(optional) => add_optional(function_index, conjunction, optional),
         typeql::Pattern::Statement(statement) => add_statement(function_index, conjunction, statement),
     })?;
+
+    conjunction
+        .conjunction
+        .nested_patterns()
+        .iter()
+        .filter_map(|nested| match nested {
+            NestedPattern::Optional(optional) => Some(optional),
+            _ => None,
+        })
+        .for_each(|optional| {
+            for var in optional.conjunction().referenced_variables() {
+                // if the variable is available in the parent scope, it's bound externally and passed in so not optional
+                if !conjunction.context.is_variable_available_in(conjunction.conjunction.scope_id(), var) {
+                    conjunction.context.set_variable_optionality(var, true)
+                }
+            }
+        });
```

Comment:
I might be able to break this.
Yes, we do it after all the constraints for this conjunction, but can we guarantee that a parent conjunction won't be adding this later?

Something like:
```
match 
{ $y isa person;  try { friend ($x, $z); friend($z, $y);  } or { $y isa page;  friend($x, $z); follows($z, $y); }
$x isa person, has id 5;
```

Reply:
In this case we can, because it's add_patterns and we can expect to add all the patterns from TypeQL at once yes - but really the architecture here isn't perfect in that we should return an immutable, finished IR at the end instead of continuing to have the builder to prevent this.